src/rust/engine/fs/store/src/lib.rs

```diff
@@ -1231,6 +1231,47 @@ impl Store {
       .boxed()
   }
 
+  ///
+  /// Returns indirect references to files in a Digest sorted by their path.
+  ///
+  pub fn entries_for_directory(
+    &self,
+    digest: Digest,
+  ) -> BoxFuture<'static, Result<Vec<FileEntry>, String>> {
+    self
+      .walk(digest, move |_, path_so_far, _, directory| {
+        future::try_join_all(
+          directory
+            .files
+            .iter()
+            .map(|file_node| {
+              let path = path_so_far.join(file_node.name.clone());
+              let is_executable = file_node.is_executable;
+              let file_node_digest = try_future!(require_digest(file_node.digest.as_ref()));
+              future::ready(Ok(FileEntry {
+                path,
+                digest: file_node_digest,
+                is_executable,
+              }))
+              .boxed()
+            })
+            .collect::<Vec<_>>(),
+        )
+        .boxed()
```

Comment:
It doesn't look like this needs to use `try_join_all`: the body of the loop over files is creating Futures unnecessarily. So maybe just imperatively create the `FileEntry`s and then have a single `future::read(Ok(fie_entries))` at the end of this closure.

Reply:
Fixed.